src/Symfony/Component/DependencyInjection/Loader/Configurator/AbstractConfigurator.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Loader\Configurator;
+
+use Symfony\Component\DependencyInjection\Argument\ArgumentInterface;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\Parameter;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\ExpressionLanguage\Expression;
+
+abstract class AbstractConfigurator
+{
+    const FACTORY = 'unknown';
+
+    protected $parent;
+    protected $definition;
+    protected $id;
+    protected $defaultTags = array();
+
+    public function __construct(AbstractConfigurator $parent, Definition $definition, $id = null, array $defaultTags = array())
+    {
+        $this->parent = $parent;
```

Comment:
I don't love that all `AbstractConfigurator` have this `AbstractConfigurator` constructor arg, because we ultimately call methods on `$this->parent` that require specific configurators to be passed. For example, a `new ServiceConfigurator()` *must* be passed a `ServicesConfigurator`, not any `AbstractConfigurator`. Plus, `$this->parent` isn't needed in this parent class. Could we just give each class its own `__construct()` method as needed (or at least remove the `$parent` from it)?